experanto/datasets.py

```diff
@@ -397,41 +397,72 @@ def initialize_transforms(self):
 
     def get_sample_in_meta_condition(self) -> dict:
         """
-        iterates through all stimuli, selects the ones which match the meta conditions (tiers or stimuli types) and creates a mask to select the correct times using `self._screen_sample_times` as the clock/reference times
+        iterates through all samples and checks if they are in the meta condition of interest
            for example:
               if meta_conditions = {"tier": [train,train, ...], "stim_type": [type1, type2, ...]}
               and valid_condition = {"tier": train, "stim_type": type2}
            then the output would be {"tier": [True, True, ...], "stim_type": [False, True, ...]}
         """
+
         sample_in_meta_condition = {}
         for k, v in self.modality_config["screen"]["valid_condition"].items():
-            sample_in_meta = []
-            for i, (condition, start, end) in enumerate(zip(self.meta_conditions[k], self._start_times, self._end_times)):
-                if condition == v or (condition == "blank" and self.modality_config["screen"]["include_blanks"]):
-                    sample_in_meta.append(
-                        (self._screen_sample_times >= start) & (self._screen_sample_times < end)
-                    )
-            sample_in_meta_condition[k] = np.stack(sample_in_meta).sum(0).astype(bool)
+            # Pre-allocate a boolean array
+            result = np.zeros_like(self._screen_sample_times, dtype=bool)
+            
+            # Create masks for all trials at once
+            valid_conditions = np.array([
+                condition == v or (condition == "blank" and self.modality_config["screen"]["include_blanks"])
+                for condition in self.meta_conditions[k]
+            ])
+            
+            # Only process valid trials
+            valid_indices = np.where(valid_conditions)[0]
+            if len(valid_indices) > 0:
+                starts = self._start_times[valid_indices]
+                ends = self._end_times[valid_indices]
+                
+                # Vectorized comparison for all valid trials
+                for start, end in zip(starts, ends):
+                    mask = (self._screen_sample_times >= start) & (self._screen_sample_times < end)
+                    result |= mask
+                    
+            sample_in_meta_condition[k] = result
+
         return sample_in_meta_condition
 
-    def get_full_valid_sample_times(self, ) -> Iterable:
+
+
+    def get_full_valid_sample_times(self) -> Iterable:
         """
-        iterates through all sample times and checks if they could be used as
-        start times, eg if the next `self.chunk_sizes["screen"]` points are still valid
-        based on the previous meta condition filtering
-        :returns:
-            valid_times: np.array of valid starting points
+        Compare original and optimized implementations
```

Comment:
Could you reinstate the "old" docstring? also feel free to extend it if you think it could be clearer